### ISSUE TITLE
Question: progressive API for network download?

### ISSUE THREAD

**Issue by SaschaNaz** [FLIF-hub/FLIF#290]
I succeeded in doing asynchronous progressive decoding when the whole image file is available but I'm not sure there is an API for partially downloaded file. I hope I can decode partial file and later decode more when more bytes are available, without restarting from the first byte. Can I do this with the exposed API set currently?

**Comment by hrj** [comment]
@SaschaNaz If your environment allows you to suspend a thread, you could suspend the decoder thread and awaken it when more data is available.

However, I would still like direct support in the API since there is no thread support in JS in the browser. JS in the browser is single threaded and there is no way to yield the thread or anything fancy.

Hence, in JS + browser, there is no way to decode partially, render an image, and then optionally resume decoding when more data is available.

**Comment by hrj** [comment]
C++ 11 seems to have `std::future`. If the IO read function returns a "future byte" instead of just a byte, we could keep most of the current code intact and yet support async io.

However, I am pretty sure emscripten doesn't handle that efficiently. They have an interpreted mode to support async calls, but that is gonna be really slow.

**Comment by jonsneyers** [comment]
What would be the best way to add this to the API?

The decoder could call the callback not just when the desired quality has been reached, but also when a premature EOF/EOS has been detected.

That would mean that we need to detect EOF/EOS early enough, so that it is possible to continue decoding. Currently the check happens not in the innermost loop (for performance reasons) but at every row beginning, which means that the premature end is actually detected when it's already too late (bogus values have been fed to the arithmetic coder), which you don't notice because the entire zoomlevel gets discarded and replaced by interpolation.

Currently the callback function has to return the desired quality for the next callback. Perhaps it should return also the max byte position to read. E.g. we could make it a void and make its arguments pointers so they can be used as output arguments. The special value "-1" (or 0 for that matter) for the max byte position could signify "no need to check the position, read as much as you want" (which would correspond to the current behavior). 

Then the decoder would need to do EOF/EOS checking sufficiently close to the inner loops to make sure it goes to the callback at the right time. Currently callbacks only happen when a full plane/zoomlevel has been decoded; this would have to be changed to allow callbacks in the middle of a plane/zoomlevel. That should be not too hard: in any case, before it calls the callback, it has to clone the buffer, interpolate whatever hasn't been decoded yet, and do all reverse transforms (e.g. YCoCg to RGB). Since it has to work on a cloned buffer anyway, it can e.g. discard the partial data on the current plane/zoomlevel, without harming the ability to continue decoding when the callback returns.

**Comment by hrj** [comment]
(I am assuming this issue is mainly for JS in browser. Most other platforms these days have good support for multi-threading.)

I am in favour of the `std::future` approach. It will require fewer changes in C++ code and is conceptually better.

The fallout for JS is:
* Emscripten port will have to use interpreter mode if it needs to support async io.
* Emscripten port can use the regular mode if it doesn't need to support async io.
* The newer versions of JS have async / await support coming up. Eventually, we can hope that emscripten can support such code directly.
* In far future, we can have a natively written JS decoder.
* In the really far future, we can hope that the JS decoder becomes obsolete and FLIF is natively supported by the browsers.

It would be good to have some JS expert chime in. I would like to summon @mrdoob (as a JS expert, I know of, who is already following the project).

**Comment by mrdoob** [comment]
`XMLHttpRequest` does have a `progress` event:

```js
var request = new XMLHttpRequest();
request.open( 'GET', url, true );
request.addEventListener( 'progress', function ( event ) {

	console.log( event );

}, false );
request.responseType = 'arraybuffer';
request.send( null );
```

Is that what you guys are looking for?

**Comment by hrj** [comment]
@mrdoob Not exactly. I mean we will have to use that XHR API yes, but the problem is more about how to marry the async event model of JS, to the blocking i/o model in C++ code.

The C++ code uses a simple function called `getChar()` to fetch the next char in the input stream. It is currently a blocking call. In JS, there is only one way I can see to implement this model: fetch the entire response, store it in a buffer, and *then* start decoding. Whenever the decoder calls `getChar()` we return a byte from the buffer, and increment the position in the buffer.

To use the XHR "progress" event, we would need to somehow introduce async i/o into the C++ code.

The alternate that @jonsneyers suggests, if I understand right, is to implement a sort of polling in the decoder: The i/o calls will remain blocking in nature. When the decoder sees that there are not enough bytes in the input stream, it will remember its state, and exit. The decoding API will be called again at a later time when more bytes are received in the input, and the decoder will try to continue from its previous state.

**Comment by SaschaNaz** [comment]
I hope I can use Fetch stream API for the future new API. I don't know about the current decoder implementation but I think something like this will be useful to me:

```cpp
// new FLIF_PARTIAL_DECODE_RESULT will store intermediate decode result 
FLIF_PARTIAL_DECODE_RESULT* partial = flif_create_partial_decode_result();
int32_t is_quality_reached = decoder->decode_partial(partial, buffer, buffer_size_bytes, target_quality);

// later
int32_t is_quality_reached = decoder->decode_partial(partial, buffer, buffer_size_bytes, target_quality);
```

**Comment by SaschaNaz** [comment]
My expected use in TypeScript (not real code):

```ts
// main thread
const response = await fetch("foo.flif");
await createArrayBufferOnWorker(response.headers.get("content-length"));
const reader = response.body.getReader();
while (true) {
  const read = reader.read();
  appendBytesOnWorker(read.value.buffer);
  if (read.done) {
    break;
  }
}

// worker:
let view: Uint8Array;
let viewPointer: number;
let writePosition: number;

function createArrayBufferHandler(size: number) {
  viewPointer = libflifem._malloc(size); // malloc on Emscripten FLIF module
  view = new Uint8Array(libflifem.buffer, viewPointer, size);
  // this will be input for decoder
}

function appendBytesHandler(buffer: ArrayBuffer) {
  view.set(new Uint8Array(buffer), writePosition); // directly write to Emscripten memory
  writePosition += buffer.byteLength;
  // TODO: request the decoder to read until `writePosition`.
  // The following line will always re-decode from the zero position
  decoder.decodeMemory(viewPointer, writePosition);
}
```

**Comment by hrj** [comment]
@SaschaNaz If I understand correctly, what you are doing there is synchronous code. Isn't it already possible with current API?

Also, I am not sure how much abstract you wanted the code sample to be, but you can't call functions in worker thread directly from main thread, nor share any memory with them (there are proposals to do that in the future). You have to pass messages between the two threads.

**Comment by hrj** [comment]
@SaschaNaz 

```
// TODO: request the decoder to read until `writePosition`.
```

This can be done with current API by writing a custom IO class that signals EOF when `writePosition` number of bytes have been read. This is exactly the approach followed by poly-flif. You can have a look at the IO code there.

**Comment by SaschaNaz** [comment]
@hrj Ah, the point there was meant to be `// The following line will always re-decode from the zero position`

This is my current implementation using Web Worker but just with opening local files. http://saschanaz.github.io/libflif.js

**Comment by hrj** [comment]
I have an idea. I don't know how feasible it is, but here goes anyway:

Eliminate callbacks and add the ability to resume decoding from a previously saved state. Essentially, we will have to manually do what the async functionality in modern languages does automatically. They store the stack frame somewhere and resume it (or something equivalent to that). We will have to store the context so that decoding can be resumed when more data is available.

Internally, the algorithm will be:
* Whenever the end of stream is reached, an exception will be thrown.
* There will be "save-points" within the decoder which will catch the exception. Just before the save-point, the decoder will store its current context.
* An opaque pointer to this context will be returned by the API, so that caller can resume decoding from this point, when more data is available.

The unknown part to me is: How much context needs to be stored, to enable resumption later.

**Comment by jonsneyers** [comment]
If the state of the memory is kept intact (the image buffers, but also the MANIAC tree, the state of the arithmetic coder, etc), then the only thing that is needed to resume decoding is basically: what's the current pixel I'm working on? (x, y, zoomlevel, channel, frame) plus a few local variables of the decode functions (https://github.com/FLIF-hub/FLIF/blob/master/src/flif-dec.cpp#L637). Of course the decoding functions would need to be adapted to allow resuming at a given point. One complication is that EOS is likely to happen in the middle of decoding some pixel value, and it gets very tricky to resume from there. If it's possible to detect an 'almost end of stream' (only 10 bytes left), then that would help a lot to simplify things. I'd need to calculate the worst-case number of bytes per pixel value to set a safe margin for that. (the worst case could be pretty bad: for 8-bit images, the number encoding we use could end up using about 20 arithmetically coded bits for one pixel value, and if the chances happen to be very unfavorable (1/4096 is the worst-case) then theoretically each of those bits could probably take a few bytes to encode.

**Comment by hrj** [comment]
How much overhead would it entail to clone the state of the MANIAC tree, arithmetic coder, etc? (I think cloning the image buffers won't be required because a pixel will be only modified when it is fully decoded).

If cloning won't have a big performance impact, then my idea is to just clone before a "save-point", say before every row (or every zoom-level, in the worst case). Then during resuming, the clone can be used and the original discarded. And then the IO code can simply throw an exception when EOS is reached.
